### PR TITLE
Sanitize proxy component name to minimize exceptions.

### PR DESCRIPTION
**Pull request by leidegre**
Sanitize the display name to be able to act as a component name without causing issues if display name is not a valid identifier. This happens all the time with Redux components.

It's a very simple test that is more aggressive than the JavaScript identifier rules but it is also very simple.

Here's a Fiddle with some tests https://jsfiddle.net/rzyfr1hz/

J

### CODE REVIEW

File: src/createClassProxy.js [hunk 1]
```diff
@@ -36,6 +36,19 @@ function getDisplayName(Component) {
...
+// This is a simple way to sanitize the component name.
+// It will prevent an exception if the component happens to use a name 
+// that is also not a valid identifier.
+const SIMPLE_IDENT_PATTERN = /^[$\w]+$/
+const SIMPLE_IDENT_FIX = /[^$\w]+/g
+
+function sanitizeFunctionName(fn) {
+  if (SIMPLE_IDENT_PATTERN.test(fn)) {
+	  return isNaN(parseInt(fn)) ? fn : '_' + fn;
+  }
```
Comment (gaearon) [inline src/createClassProxy.js]: Nitpick: too many spaces before return.

**Comment by gaearon** [comment]
Thanks! Can you please add some tests? I think displayName tests are in consistency.js.

**Comment by leidegre** [comment]
I'll fix that together with the spaces later today.

**Comment by leidegre** [comment]
I took a look at the things you mentioned.

- The code is covered by existing tests (not every edge case).
- There's no way to tell/assert (in it's current form whether these exceptions were thrown or not)

I'd much rather like to provide the function as a utility and tests that in isolation. Just to ensure that its behavior is as it should. I'm not overly familiar with the source code but from what I can tell though, the component name is overshadowed anyway.

~~~
try {
  Object.defineProperty(ProxyComponent, 'name', {
    value: displayName
  });
} catch (err) { }
~~~

Unless I'm missing something. This is only really necessary if we can't create a dynamic constructor and `defineProperty` `'name'` does not work which seems like the most spectacular edge case. At this point, do we really care that we didn't set a component name for the proxy? Wouldn't just be simpler to do this.

~~~
let displayName = getDisplayName(InitialComponent);
ProxyComponent = function () {
  return instantiate(() => CurrentComponent, this, arguments);
};
try {
  Object.defineProperty(ProxyComponent, 'name', {
    value: displayName
  });
} catch (err) {
  // ...
}
~~~

Please correct me if I'm off here.

**Comment by gaearon** [comment]
> if we can't create a dynamic constructor

We have [a test](https://github.com/gaearon/react-proxy/blob/next/test/consistency.js#L301) for this. Let’s just give that function a weird name.

**Comment by leidegre** [comment]
@gaearon like this https://github.com/gaearon/react-proxy/pull/59 ?
